### PR TITLE
Integrate API: [GET] /api/homework #136,Homework List page #135

### PR DESCRIPTION
**Pull request by timt1028**
resolve: #135 #136  

截圖 2021-06-24 上午12 15 11

### CODE REVIEW

File: src/app/App.js [hunk 1]
```diff
@@ -26,7 +28,9 @@ function App() {
                         <PrivateRoute path="/problems" component={ProblemList}/>
                         <PrivateRoute path="/exams" component={ExamList}/>
                         <PrivateRoute path="/students" component={StudentList}/>
-                        <PrivateRoute path="/groups" component={GroupList}/>
+                        <PrivateRoute path="/homework" component={HomeworkList}/>
+                        <PrivateRoute path="/groups" exact component={GroupList}/>
+                        <PrivateRoute path="/groups/:groupId/students" component={GroupMembers}/>
```
Comment (Wally5077) [inline src/app/App.js:33]: 這個是？

File: src/app/homework/HomeworkList.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    const refetchHomework = useCallback(() => {
+        homeworkService.getAllHomework()
+            .then(homeworkList => {
+                setHomeworkList(homeworkList);
+            })
+    }, [setHomeworkList]);
+
+    useEffect(() => {
+        if (!homeworkList || homeworkList.length === 0) {
+            refetchHomework()
```
Comment (Wally5077) [inline src/app/homework/HomeworkList.js:26]: 分號

File: src/app/homework/HomeworkList.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        }
+    }, [homeworkList, refetchHomework]);
+
+    const onHomeworkCreated = () => {
+    };
```
Comment (Wally5077) [inline src/app/homework/HomeworkList.js:31]: 加上 todo 標明下個 issue 會完成

File: src/app/homework/HomeworkList.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                                  <TableCell>#</TableCell>,
+                                  <TableCell>Homework Name</TableCell>,
+                                  <TableCell>Problem IDS</TableCell>,
+
+                                  <EmptyCell/>
+                              ]}
+                              tableRowGenerator={{
+                                  list: homeworkList,
+                                  key: (homework) => homework.id,
+                                  data: (homework) => [
```
Comment (Wally5077) [inline src/app/homework/HomeworkList.js:67]: homework 拿掉

File: src/app/homework/modals/CreateHomeworkModal.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import '../../commons/modals/TextareaModal.scss'
+import {createRef, useState} from "react";
+import {renderModal} from "../../commons/modals/modal";
+import {ModalInput} from "../../commons/modals/ModalInput";
+import "../../homework/modals/CreateHomeworkModel.css";
+import {ModalHeader} from "../../commons/modals/ModalHeader";
+
+
+const CreateHomeworkModal = ({show, onClose}) => {
+    const [name, setName] = useState(null);
```
Comment (Wally5077) [inline src/app/homework/modals/CreateHomeworkModal.js:11]: const [name, setName] = useState(undefined);

File: src/app/homework/modals/CreateHomeworkModal.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    const nameInputRef = createRef();
+    const closeIconRef = createRef();
+
+    const handleSubmit = e => {
+        e.preventDefault();
+    };
+
+    let formRef;
+
+    const [content, setContent] = useState();
```
Comment (Wally5077) [inline src/app/homework/modals/CreateHomeworkModal.js:21]: 變數移上去

File: src/services/HomeworkService.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    async createHomework({name, problemIds}) {
+        return this.axios.post('/api/homework', new Homework({name, problemIds}))
+            .then(res => new Homework(res.data));
+    }
+
+    async deleteEHomework(homeworkId) {
+        return this.axios.delete(`/api/homework/${homeworkId}`)
+            .then(res => res);
+    }
```
Comment (Wally5077) [inline src/services/HomeworkService.js:32]:
```Js
    async deleteHomework(homeworkId) {
        return this.axios.delete(`/api/homework/${homeworkId}`);
    }
```

File: src/services/HomeworkService.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    constructor(studentService) {
+        super({
+            baseURL: process.env.REACT_APP_HOMEWORK_SVC_BASE_URL,
+            timeout: 10000,
+            tokenSupplier: studentService.currentToken
+        });
+    }
+
+    async getHomework(homeworkId) {
```
Comment (Wally5077) [inline src/services/HomeworkService.js:14]: getHomeworkById

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: .env [hunk 9]
```diff
@@ -1,4 +1,5 @@
 REACT_APP_STUDENT_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw
 REACT_APP_EXAM_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw
 REACT_APP_PROBLEM_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw
-REACT_APP_SUBMISSION_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw
\ No newline at end of file
+REACT_APP_SUBMISSION_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw
+REACT_APP_HOMEWORK_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw
```
Comment (Johnny850807) [inline .env]: Reduce `REACT_APP_EXAM_SVC_BASE_URL` and `REACT_APP_HOMEWORK_SVC_BASE_URL` into 
`REACT_APP_ACADEMY_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw`

remember to also change the services which depend on the exam or homework service's link

File: src/app/App.js (hunk 1 above)
Comment (Johnny850807) [reply src/app/App.js:33]: same, ... what is this?

File: src/app/homework/HomeworkList.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                                      <TableCell>{homework.problemIds.join(', ')}</TableCell>,
+                                      <TableCell>{actionItemsButton({homework: homework})}</TableCell>,
+                                  ]
+                              }}/>
+
+                <CreateHomeworkModal show={showCreateHomeworkModal}
+                                     onClose={() => setShowCreateHomeworkModal(false)}
+                                     onHomeworkCreated={onHomeworkCreated}/>
+
+                <RemoveConfirmationModal title={"Delete this homework"}
```
Comment (Johnny850807) [inline src/app/homework/HomeworkList.js:79]: rename `RemoveConfirmationModal` to `DeleteConfirmationModal`

File: src/app/homework/HomeworkList.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+import FakeLink from "../commons/FakeLink";
+import {RemoveConfirmationModal} from "../commons/modals/RemoveConfirmationModal";
+import {CreateButton} from "../commons/buttons/CreateButton";
+import {CreateHomeworkModal} from "./modals/CreateHomeworkModal";
+
+const HomeworkList = () => {
+    const [showCreateHomeworkModal, setShowCreateHomeworkModal] = useState(false);
+    const [homeworkList, setHomeworkList] = useState(undefined);
+    const [selectedHomework, setSelectedHomework] = useState(undefined);
+    const [showRemoveHomeworkConfirmationModal, setShowRemoveHomeworkConfirmationModal] = useState(false);
```
Comment (Johnny850807) [inline src/app/homework/HomeworkList.js:16]: rename `Remove -> Delete`

File: src/app/homework/HomeworkList.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    };
+
+    if (!homeworkList) {
+        return <Spinner/>
+    }
+
+    const actionItemsButton = ({homework}) =>
+        <ThreeDotsButton dropDownItems={[
+            {
+                name: "Remove",
```
Comment (Johnny850807) [inline src/app/homework/HomeworkList.js:40]: Remove -> Delete

File: src/app/homework/HomeworkList.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            <div style={{display: "flex", justifyContent: "center"}}>
+                <ItemListPage width="1200px"
+                              title="Homework"
+                              filterItems={["Filter", "Homework Name", "Problem IDS"]}
+                              Button={() =>
+                                  <CreateButton onClick={() => setShowCreateHomeworkModal(true)}/>}
+                              tableHeaders={[
+                                  <TableCell>#</TableCell>,
+                                  <TableCell>Homework Name</TableCell>,
+                                  <TableCell>Problem IDS</TableCell>,
```
Comment (Johnny850807) [inline src/app/homework/HomeworkList.js:60]: Problem IDs

File: src/app/homework/HomeworkList.js [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                }
+            },
+        ]}/>
+
+    return (
+        <div className="font-poppins" style={{paddingTop: "20px", paddingBottom: "150px"}}>
+            <div style={{display: "flex", justifyContent: "center"}}>
+                <ItemListPage width="1200px"
+                              title="Homework"
+                              filterItems={["Filter", "Homework Name", "Problem IDS"]}
```
Comment (Johnny850807) [inline src/app/homework/HomeworkList.js:54]: Problem IDs

File: src/app/homework/HomeworkList.js [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+                                  <EmptyCell/>
+                              ]}
+                              tableRowGenerator={{
+                                  list: homeworkList,
+                                  key: (homework) => homework.id,
+                                  data: (homework) => [
+                                      <TableCell>{homework.id}</TableCell>,
+                                      <FakeLink>{homework.name}</FakeLink>,
+                                      <TableCell>{homework.problemIds.join(', ')}</TableCell>,
+                                      <TableCell>{actionItemsButton({homework: homework})}</TableCell>,
```
Comment (Johnny850807) [inline src/app/homework/HomeworkList.js:71]: `{homework: homework}` -> `{homework}`

File: src/app/homework/modals/CreateHomeworkModel.css [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+.create-homework-modal .schedule-item * {
+    line-height: 25px;
+    height: 25px;
+}
+
+.create-homework-modal .dot {
+    height: 10px;
+    margin-top: auto;
+    margin-bottom: auto;
+}
```
Comment (Johnny850807) [inline src/app/homework/modals/CreateHomeworkModel.css:18]: use `scss` syntac to nest the styles
1. change the file's extension name from `.css` to `.scss`
2. Nest the styles as belows
```scss
.create-homework-modal {
    #create-btn {
         ..
    }
    .dot {
       ...
    }
}
```

File: src/services/HomeworkService.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import AbstractService from "./AbstractService";
+import Homework from "../models/Homework";
+
+export class HomeworkService extends AbstractService {
+
+    constructor(studentService) {
+        super({
+            baseURL: process.env.REACT_APP_HOMEWORK_SVC_BASE_URL,
```
Comment (Johnny850807) [inline src/services/HomeworkService.js:8]: changed to `REACT_APP_ACADEMY_SVC_BASE_URL`

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done! :D

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Review Done!
By the way, can you take the screenshot of `createHomework modal` and `deleteHomework modal`?
So that I can give some comments to UI. Thanks. 👍🏻

File: src/app/App.js (hunk 1 above)
Comment (nyngwang) [reply src/app/App.js:33]: `/group`的子項要放到已經做好的`GroupList`(line 60~70)底下哦

File: src/app/homework/modals/CreateHomeworkModal.js (hunk 5 above)
Comment (nyngwang) [reply src/app/homework/modals/CreateHomeworkModal.js:11]: Prefer `undefined` than `null`, the former is the default init value.

**Review by nyngwang — CHANGES_REQUESTED** [review]
Some comment appended.

File: src/app/homework/modals/CreateHomeworkModel.css (hunk 16 above)
[the remark this answers is not in the document]
Comment (timt1028) [reply src/app/homework/modals/CreateHomeworkModel.css:18]: Ok  thanks , 
And the screenshot of createHomework modal and deleteHomework modal in the another issue.
 #139 https://github.com/Judge-Girl/Judge-Girl-CMS/issues/139

File: .env (hunk 9 above)
Comment (Johnny850807) [reply .env]: should also modify the exam service

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

File: src/app/App.js (hunk 1 above)
Comment (Wally5077) [reply src/app/App.js]: 我留言是 rebase 問題，這個沒 rebase 好。

File: src/services/HomeworkService.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+        return this.axios.get(`/api/homework`)
+            .then(res => res.data.map(obj => new Homework(obj)));
+    }
+
+    async createHomework({name, problemIds}) {
+        return this.axios.post('/api/homework', new Homework({name, problemIds}))
+            .then(res => new Homework(res.data));
+    }
+
+    async deleteEHomework(homeworkId) {
```
Comment (Wally5077) [inline src/services/HomeworkService.js:29]: 名字還是有問題

File: src/app/homework/modals/CreateHomeworkModal.js [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+import {renderModal} from "../../commons/modals/modal";
+import {ModalInput} from "../../commons/modals/ModalInput";
+import {ModalHeader} from "../../commons/modals/ModalHeader";
+
+
+const CreateHomeworkModal = ({show, onClose}) => {
+    const [name, setName] = useState(undefined);
+    const nameInputRef = createRef();
+    const closeIconRef = createRef();
+    const [content, setContent] = useState();
```
Comment (Wally5077) [inline src/app/homework/modals/CreateHomeworkModal.js:13]: useState 請放一起

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by nyngwang — CHANGES_REQUESTED** [review]
Review Done.

**Review by Chaoyuuu — APPROVED** [review]
LGTM

**Review by Wally5077 — APPROVED** [review]
LGTM

**Review by nyngwang — APPROVED** [review]
LGTM.

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Pull request merged**